src/main.cpp

```diff
@@ -3627,6 +3652,29 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
         // while still invalidating it.
         if (mutated)
             return state.DoS(100, false, REJECT_INVALID, "bad-txns-duplicate", true, "duplicate transaction");
+
+        // Merge mining checks.
+        if (!block.m_aux_pow.IsNull()) {
+            // Check that auxiliary proof-of-work data have canonical encoding.
+            auto aux_hash = block.GetAuxiliaryHash(&mutated);
+            if (mutated) {
+                return state.DoS(100, false, REJECT_INVALID, "bad-auxpow-mutated", true, "auxiliary proof-of-work header is non-canonical");
+            }
+
+            // The auxiliary proof-of-work is committed to in the coinbase string.
+            if (block.vtx.empty() || block.vtx[0].vin.empty() || (block.vtx[0].vin[0].scriptSig.size() < 32) || memcmp(aux_hash.second.begin(), &(block.vtx[0].vin[0].scriptSig.end()-32)[0], 32)) {
+                return state.DoS(100, false, REJECT_INVALID, "bad-auxpow-commit", true, "incorrect commitment to auxiliary proof-of-work in coinbase string");
+            }
+
+            // Check that CAuxProofOfWork::m_commit_hash_merkle_root is correct.
+            CMutableTransaction cb(block.vtx[0]);
+            cb.vin[0].scriptSig = CScript();
+            cb.vin[0].nSequence = 0;
```

Comment:
In Tradecraft, `scriptSig` and `nSequence` of the coinbase input are cleared for the segwit commitment, which allows the miner is allowed alter these fields (e.g. incrementing the extranonce) without having to recalculate the segwit commitment. (Aside: originally I was going to have just nSequence be cleared, because 4 bytes is enough extranonce for everybody, right? But it turns out that a separate soft-fork which requires the coinbase's `nLockTime` field to be set to the median-time-past causes a non-final `nSequence` value to invalidate the block. Oops.)

Anyway the point is that these fields may have been altered by the miner who later solved the native proof-of-work, so they can't be included in this commitment.

Reply:
Ahh extra nonce space. Got it!

Reply:
I would maybe add `// not in commitment as miner may use this later` in.

Reply:
Added comment.